src/CAN/CANMotor.h

```diff
@@ -0,0 +1,158 @@
+ *
+ * These motor boards can either be AVR or PSoC boards.
+ */
+namespace can::motor {
+
+/** @brief The possible motor modes. */
+enum class motormode_t { pwm = MOTOR_UNIT_MODE_PWM, pid = MOTOR_UNIT_MODE_PID };
+
+/** @brief The supported motor position sensors. */
+struct sensor_t {
+	enum { encoder = 0, potentiometer = 1 };
+};
+
+/**
+ * @brief Emergency stop all motors on the CAN bus.
+ *
+ * This cuts power to all motors and resets them.
+ */
+void emergencyStopMotors();
+
+/**
+ * @brief Initialize a motor.
+ *
+ * This does not initialize any sensor or controller.
+ *
+ * @param serial The CAN serial number of the motor to initialize.
+ */
+void initMotor(deviceserial_t serial);
+
+/**
+ * @brief Initialize an encoder attached to the given motor.
+ *
+ * For potentiometers, use initPotentiometer().
+ *
+ * @param serial The CAN serial number of the motor board.
+ * @param invertEncoder If true, invert the encoder direction. Use this to correct sensor
+ * phase.
+ * @param zeroEncoder If true, reset the encoder position to zero.
+ * @param pulsesPerJointRev The number of encoder pulses per revolution of the physical joint.
+ * Measure/calculate this using the gear ratios and encoder specs.
+ * @param telemetryPeriod An optional parameter specifying the telemetry period.
+ * The telemetry will be fetched at this period automatically. An empty optional disables this
+ * behavior, in which case the motor position must be explicitly pulled.
+ */
+void initEncoder(deviceserial_t serial, bool invertEncoder, bool zeroEncoder,
+				 int32_t pulsesPerJointRev,
+				 std::optional<std::chrono::milliseconds> telemetryPeriod);
+
+// TODO: figure out the correct parameters needed and implement for both motor boards
+void initPotentiometer(deviceserial_t serial, int32_t posLo, int32_t posHi,
+					   std::optional<std::chrono::milliseconds> telemetryPeriod);
+
+/**
+ * @brief Set the PID constants for a motor board.
+ *
+ * Note that the PID constants are specified in units of 10000, so a 1 is interpreted as a
+ * 10000. This is because the controller operates on millidegrees.
+ *
+ * @param serial The CAN serial number of the motor board.
+ * @param kP The P coefficient.
+ * @param kI The I coefficient.
+ * @param kD The D coefficient.
+ */
+void setMotorPIDConstants(deviceserial_t serial, int32_t kP, int32_t kI, int32_t kD);
+
+/**
+ * @brief Set the mode of a motor board.
+ *
+ * @param serial The CAN serial number of the motor board.
+ * @param mode The mode to set.
+ */
+void setMotorMode(deviceserial_t serial, motormode_t mode);
+
+/**
+ * @brief Set the power output of a motor board.
+ *
+ * @param serial The CAN serial number of the motor board.
+ * @param power Percent power, in the range [-1,1].
+ */
+void setMotorPower(deviceserial_t serial, double power);
+
+/**
+ * @brief Set the power output of a motor board.
+ *
+ * The motor mode should have been set to motormode_t::pwm.
```

Comment:
I left that as undefined behavior, since I think it's probably reasonable to expect users to put the motor in the right mode. Sending a modeset packet every time would be expensive, and I didn't want to track every motor's mode, since we don't necessarily know the serial of every motor.

So the easiest solution is to just have users set the modes of the motor correctly, which I think makes sense.